packages/core/src/compiler/Style.ts

```diff
@@ -1162,47 +1162,67 @@ const matchStyArgToSubArg = (
       subArg
     );
   }
-  return undefined;
+  return [];
 };
 
 /**
  * Match a list of Style arguments against a list of Substance arguments.
- * @returns If all arguments match, return a `Subst` that maps the Style variable(s) against Substance variable(s). If any arguments fail to match, return `undefined`.
+ * @returns If all arguments match, return a `Subst[]` that contains mappings which map the Style variable(s) against Substance variable(s). If any arguments fail to match, return [].
  */
 const matchStyArgsToSubArgs = (
   styTypeMap: { [k: string]: StyT<A> },
   subTypeMap: { [k: string]: TypeConsApp<A> },
   varEnv: Env,
   styArgs: PredArg<A>[] | SelExpr<A>[],
   subArgs: SubPredArg<A>[] | SubExpr<A>[]
-): Subst | undefined => {
-  const initRSubst: Subst | undefined = {};
-  const res = zip2<PredArg<A> | SelExpr<A>, SubPredArg<A> | SubExpr<A>>(
-    styArgs,
-    subArgs
-  ).reduce((rSubst: Subst | undefined, [styArg, subArg]) => {
-    if (rSubst === undefined) {
-      return undefined;
-    }
-    const argSubst = matchStyArgToSubArg(
+): Subst[] => {
+  const stySubArgPairs = zip2<
+    PredArg<A> | SelExpr<A>,
+    SubPredArg<A> | SubExpr<A>
+  >(styArgs, subArgs);
+
+  const substsForEachArg = stySubArgPairs.map(([styArg, subArg]) => {
+    const argSubsts = matchStyArgToSubArg(
       styTypeMap,
       subTypeMap,
       varEnv,
       styArg,
       subArg
     );
-    if (argSubst === undefined) {
-      return undefined;
-    } else {
-      return { ...rSubst, ...argSubst };
-    }
-  }, initRSubst);
-  return res;
+    return argSubsts;
+  });
+
+  // We do Cartesian product here.
```

Comment:
Not gonna improve the computational complexity but maybe good for space complexity: instead of performing the Cartesian product right away, let `cartesianProduct` take in a lambda to filter out the inconsistent substs. My relational algebra is super rusty, but if the condition is `consistentSubsts` is this a... natural join??

Reply:
Hmmm yeah that's something I can do

Reply:
Do you mind adding the filter lambda in? I think this will future-proof us from memory problems down the road even if we don't end up optimizing the matching algorithm.

Reply:
Oops forgot to do that - a new push will be imminent, sorry